Mathlib/Topology/UniformSpace/UniformConvergence.lean

```diff
@@ -239,6 +239,14 @@ theorem UniformContinuous.comp_tendstoUniformly [UniformSpace γ] {g : β → γ
     (hg : UniformContinuous g) (h : TendstoUniformly F f p) :
     TendstoUniformly (fun i => g ∘ F i) (g ∘ f) p := fun _u hu => h _ (hg hu)
 
+/-- Composing on the left by a uniformly continuous on the codomain function preserves uniform
+  convergence -/
+theorem UniformContinuousOn.comp_tendstoUniformly (s : Set β) (F : ι → γ → s) (f : γ → s)
```

Comment:
Please use `{s : Set β} {F : ι → γ → β} {f : γ → β}` and assume `∀ i x, F i x ∈ s` and `∀ x, f x ∈ s`.

Reply:
Sorry, I'm being a bit stupid, is there an easy way to prove it using this formulation? The old proof breaks, and I'm struggling to get it to work.

Reply:
I think the reason this is true is that `Subtype.val : ↥s → β` is a `UniformEmbedding`. Here's a proof (which could be cleaned up a bit if we had any lemmas relating `UniformEmbedding` and `TendstoUniformly`, but we don't), but it doesn't work right here because this file doesn't know about `UniformEmbedding` :sad:, which makes sense.
```
import Mathlib

open Topology Uniformity Filter Set

universe u v w x

variable {α : Type u} {β : Type v} {γ : Type w} {ι : Type x} [UniformSpace β] [UniformSpace γ] {F : ι → α → β} {f : α → β} {s s' : Set α}
  {x : α} {p : Filter ι} {p' : Filter α} {g : ι → α}

/-- Composing on the left by a uniformly continuous on the codomain function preserves uniform
  convergence -/
theorem UniformContinuousOn.comp_tendstoUniformly' (s : Set β) (F : ι → α → β) (f : α → β)
    (hF : ∀ i x, F i x ∈ s) (hf : ∀ x, f x ∈ s)
    {g : β → γ} (hg : UniformContinuousOn g s) (h : TendstoUniformly F f p) :
    TendstoUniformly (fun i => fun x =>  g  (F i x)) (fun x => g (f x)) p := by
  rw [uniformContinuousOn_iff_restrict] at hg
  lift F to ι → α → s using hF with F' hF'
  lift f to α → s using hf with f' hf'
  rw [tendstoUniformly_iff_tendsto] at h
  have : Tendsto (fun q : ι × α ↦ (f' q.2, (F' q.1 q.2))) (p ×ˢ ⊤) (𝓤 s) :=
    h.of_tendsto_comp uniformEmbedding_subtype_val.comap_uniformity.le
  apply (UniformContinuous.comp_tendstoUniformly hg ?_)
  rw [← tendstoUniformly_iff_tendsto] at this
```
Also, it's not clear to me why you are restricting to `g : β → β`, so I changed the type signature to better match other things in this file, so `f : α → β` and `g : β → γ`.

Reply:
Ok nice, this proof works in the `UniformConvergenceTopology` file so I put it there. I don't think I meant to restrict `g` like that, but I agree this is better, thanks!